modules/host/update.go

```diff
@@ -313,7 +295,25 @@ func (h *Host) ProcessConsensusChange(cc modules.ConsensusChange) {
 		h.log.Println(err)
 	}
 
-	// Handle the list of action items.
+	// Handle the list of action items. Action items require host locks and
+	// potentially require waiting for long periods of time while various
+	// network communications finish. To prevent the host lock from being held
+	// while the network I/O is blocking the action item execution, the action
+	// items are handled in their own thread. A wait group is used to prevent
+	// the host from releasing ThreadGroup control until all threads that this
+	// function have spawned have completed.
+	wg := new(sync.WaitGroup)
+	// The host should not release thread group control until all spawned
+	// threads have completed execution, though waiting for them should not
+	// block this function. The host should also not release thread group
+	// control until this function has terminated, hence the use of 'defer
+	// func() { go func() }' to coordinate waiting and done-ing.
+	defer func() {
+		go func() {
+			wg.Wait()
```

Comment:
I think there are a couple modules that want to avoid blocking ProcessConsensusChange. Usually the problem is that later blocks may invalidate previous blocks (e.g. the contractor trying to renew contracts before being fully synced).

Different concurrency strategies are appropriate for different problems. When you're just using concurrency to speed things up, the parent/child approach makes a lot of sense. But a lot of our code treats goroutines as independent actors that talk to each other. We should read up on the Smalltalk/Erlang approach to concurrency. I bet there's some good ideas to be found there.